Add BufferDict mirroring ParameterDict behavior

Balandat opened the pull request:
Summary:
Adds a `torch.nn.BufferDict` container that does the obvious thing, as requested in #37386.

Implementation-wise, there is a lot of repeated code, and it would probably make sense to write sth like a `TensorDict` base class and then subclass both `ParameterDict` and `BufferDict` from it, but for now this replicates the logic.

Test Plan: Unit tests

Differential Revision: D21270932

zdevito reviewed (commented):
I think adding this is fine, but I agree that we shouldn't just copy/paste the code of ParameterDict to change a few functions. I am pretty sure this can be done by just having a few abstract functions that access `_parameter`  or `_buffers` and call `register_x`.

Balandat commented:
Cool, I'll polish that up then.

Balandat committed:
Add BufferDict mirroring ParameterDict behavior (#37385)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/37385

Adds a `torch.nn.BufferDict` container that does the obivous thing.

In order to avoid toms of repeated code, this introduces a `_DictContainer` abstract base class to be mixed in for the `ParameterDict` and `BufferDict` implementations.

The docstrings here still need updating. The errors/messages are dynamically generated from the class name, but the docstrings are from the `ParameterDict`. We could think about using some dynamic generation of the docstrings depending on the class, but I wanted to first get feedback on the basic design here.

Test Plan: Unit tests

Differential Revision: D21270932

fbshipit-source-id: e51e8ff5ef6dac9c442cdd21164fab48fa64a9f7

Balandat commented:
In order to avoid tons of repeated code, I changed this to introduce a `_DictContainer` abstract base class that is mixed in for the `ParameterDict` and `BufferDict` implementations, which makes those very brief.

While the errors/messages are dynamically generated from the class name, the docstrings here are from the `ParameterDict` and still need updating.. We could think about using some dynamic way of generating of the docstrings depending on the class using a decorator, but I wanted to first get feedback on the design before considering that.
